packages/api/models/pdrs.js

```diff
@@ -100,24 +100,44 @@ class Pdr extends Manager {
     };
 
     record.duration = (record.timestamp - record.createdAt) / 1000;
+    this.constructor.recordIsValid(record, this.schema);
     return record;
   }
 
   /**
-   * Create a new PDR record from incoming SNS messages
+   * Try to update a PDR record from a cloudwatch event.
+   * If the record already exists, only update if the execution is different (re-run case).
    *
-   * @param {Object} payload - SNS message containing the output of a Cumulus Step Function
-   * @returns {Promise<Object>} a PDR record
+   * @param {Object} cumulusMessage - cumulus message object
    */
-  createPdrFromSns(payload) {
-    const pdrObj = get(payload, 'payload.pdr', get(payload, 'meta.pdr'));
-    const pdrName = get(pdrObj, 'name');
-
-    if (!pdrName) return Promise.resolve();
-
-    const pdrRecord = Pdr.generatePdrRecord(payload);
+  async storePdrsFromCumulusMessage(cumulusMessage) {
+    const pdrRecord = this.generatePdrRecord(cumulusMessage);
+    if (!pdrRecord) return null;
+    const updateParams = await this.generatePdrUpdateParamsFromRecord(pdrRecord);
+    if (pdrRecord.status === 'running') {
+      updateParams.ConditionExpression = 'execution <> :execution OR progress < :progress';
```

Comment:
So only update if the execution is different (re-run case) or the progress currently stored is **less than the value we're trying to update it to**

Reply:
Correct. 💯